common/src/lib.rs

```diff
@@ -0,0 +1,164 @@
+    async fn new(
+        public_key: PublicKey,
+        private_key: PrivateKey,
+        known_ids: Vec<PublicKey>,
+        known_peers: Vec<std::net::Ipv4Addr>,
+        network_id: String,
+    ) -> Result<Self, String>
+    where
+        Self: Sized;
+    /// Broadcasts a message to the network, after signed by the key given to this instance.
+    async fn broadcast(&self, message: &[u8]) -> Result<(), String>;
+    /// Creates a receiver for every message broadcasted to the network, except the one sent by this instance.
+    async fn create_recv_queue(&self) -> Result<mpsc::Receiver<Vec<u8>>, ()>;
+    /// Provides the estimated lists of live nodes that are eligible and identified by their public keys.
+    async fn get_live_list(&self) -> Result<Vec<PublicKey>, ()>;
+}
+
+/// TODO: Provide error types.
+///
+/// Note: This trait is quite subject to change.
+#[async_trait]
+pub trait UnauthorizedNetwork {
+    /// Joins the network with an authorized identity.
+    async fn new(known_peers: Vec<std::net::Ipv4Addr>, network_id: String) -> Result<Self, String>
+    where
+        Self: Sized;
+    /// Broadcasts a message to the network.
+    async fn broadcast(&self, message: &[u8]) -> Result<(), String>;
+    /// Creates a receiver for every message broadcasted to the network, except the one sent by this instance.
+    async fn create_recv_queue(&self) -> Result<mpsc::Receiver<Vec<u8>>, ()>;
+    /// Provides the estimated lists of live nodes identified by their IP addresses
+    async fn get_live_list(&self) -> Result<Vec<std::net::Ipv4Addr>, ()>;
+}
+
+/// TODO: Provide error types.
+///
+/// Note: This trait is quite subject to change.
+#[async_trait]
+pub trait KVStore {
+    /// Creates an empty store with the path to newly create.
+    async fn new(path: &str) -> Result<Self, ()>
+    where
+        Self: Sized;
+    /// Open an existing store with the path given.
+    async fn open(path: &str) -> Result<Self, ()>
+    where
+        Self: Sized;
+    /// Records the current state to the persistent storage.
+    async fn commit_checkpoint(&mut self) -> Result<(), ()>;
+    /// Reverts all the changes made since the last checkpoint.
+    async fn revert_to_latest_checkpoint(&mut self) -> Result<(), ()>;
+    /// Inserts a key-value pair into the store. If exists, it will be overwritten.
+    async fn insert_or_update(&mut self, key: Hash256, value: &[u8]) -> Result<(), ()>;
+    /// Removes a key-value pair from the store. If not exists, it will fail.
+    async fn remove(&mut self, key: Hash256) -> Result<(), ()>;
+    /// Retrieves the value associated with the key. If not exists, it will return `None`.
+    async fn get(&self, key: Hash256) -> Result<Option<Vec<u8>>, ()>;
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct Header {
+    /// The author of this block.
+    pub author: PublicKey,
+    /// The signature of the previous block.
+    pub prev_signature: Vec<Signature>,
```

Comment:
`prev_signature` -> `prev_precommit_signatures`?

I'm not sure if it would be a good idea to use the same `Signature` type for signatures for precommits and transactions.

Reply:
1. The term `precommit` sounds quite specific to some particular consensus algorithms
2. What would be the potential problem of using the same signing scheme for both transactions and consensus messages?

Reply:
1. Okay, but I think `signature` is too abstract. It's better to specify what the signature is for (perhaps `prev_block_finialization_proof`, if it is not too long)
And it should be `signatures`, not `signature`.

2. I'm okay with using the same signature scheme for the two cases, but I think it might be better to distinguish them in the type level, because checking the signature of a transaction against the content of a consensus message (or vice versa) is always a bug.